tests/samplers_tests/test_nsga2.py

```diff
@@ -477,3 +477,302 @@ def test_call_after_trial_of_random_sampler() -> None:
+        )
+
+
+def test_undxm() -> None:
+    n_trials = 4
+    n_objectives = 1
+
+    sampler = NSGAIISampler(population_size=4, crossover_name="undxm")
+
+    study = optuna.create_study(directions=["minimize"] * n_objectives, sampler=sampler)
+    study.optimize(
+        lambda t: [t.suggest_float(f"x{i}", 0, 1) for i in range(n_objectives)], n_trials=n_trials
+    )
+
+    assert len(study.trials) == n_trials
+
+    n_trials = 4
+    n_objectives = 2
+
+    sampler = NSGAIISampler(population_size=4, crossover_name="undxm")
+
+    study = optuna.create_study(directions=["minimize"] * n_objectives, sampler=sampler)
+    study.optimize(
+        lambda t: [t.suggest_float(f"x{i}", 0, 1) for i in range(n_objectives)], n_trials=n_trials
+    )
+
+    assert len(study.trials) == n_trials
+
+    n_trials = 4
+    n_objectives = 3
+
+    sampler = NSGAIISampler(population_size=4, crossover_name="undxm")
+
+    study = optuna.create_study(directions=["minimize"] * n_objectives, sampler=sampler)
+    study.optimize(
+        lambda t: [t.suggest_float(f"x{i}", 0, 1) for i in range(n_objectives)], n_trials=n_trials
+    )
+
+    assert len(study.trials) == n_trials
+
+    with pytest.raises(ValueError):
+        sampler = NSGAIISampler(population_size=2, crossover_name="undxm")
+        study = optuna.create_study(directions=["minimize"] * n_objectives, sampler=sampler)
+        study.optimize(
+            lambda t: [t.suggest_float(f"x{i}", 0, 1) for i in range(n_objectives)],
+            n_trials=n_trials,
+        )
+
+
+def test_spx() -> None:
+    n_trials = 4
+    n_objectives = 1
+
+    sampler = NSGAIISampler(population_size=3, crossover_name="spx")
+
+    study = optuna.create_study(directions=["minimize"] * n_objectives, sampler=sampler)
+    study.optimize(
+        lambda t: [t.suggest_float(f"x{i}", 0, 1) for i in range(n_objectives)], n_trials=n_trials
+    )
+
+    n_trials = 4
+    n_objectives = 2
+
+    sampler = NSGAIISampler(population_size=3, crossover_name="spx")
+
+    study = optuna.create_study(directions=["minimize"] * n_objectives, sampler=sampler)
+    study.optimize(
+        lambda t: [t.suggest_float(f"x{i}", 0, 1) for i in range(n_objectives)], n_trials=n_trials
+    )
+
+    n_trials = 4
+    n_objectives = 3
+
+    sampler = NSGAIISampler(population_size=3, crossover_name="spx")
+
+    study = optuna.create_study(directions=["minimize"] * n_objectives, sampler=sampler)
+    study.optimize(
+        lambda t: [t.suggest_float(f"x{i}", 0, 1) for i in range(n_objectives)], n_trials=n_trials
+    )
+
+    assert len(study.trials) == n_trials
+
+    with pytest.raises(ValueError):
+        sampler = NSGAIISampler(population_size=2, crossover_name="spx")
+        study = optuna.create_study(directions=["minimize"] * n_objectives, sampler=sampler)
+        study.optimize(
+            lambda t: [t.suggest_float(f"x{i}", 0, 1) for i in range(n_objectives)],
+            n_trials=n_trials,
+        )
```

Comment:
How about using `pytest.mark.parametrize` to reduce the number of codes?

Reply:
Tests here are very general besides the negative exception checks (which are great). I think they're good to include in terms of code coverage but we could leave a TODO to move them to a more general test file such as `test_samplers.py` (there are actually two files named as the such, so one of them) to reduce test code.